packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -139,33 +139,16 @@ export class AssociationManager {
     }
   }
 
-
   public static associateTagOptions(resource: cdk.IResource, resourceId: string, tagOptions: TagOptions): void {
-    const resourceStack = cdk.Stack.of(resource);
-    for (const [key, tagOptionsList] of Object.entries(tagOptions.tagOptionsMap)) {
-      InputValidator.validateLength(resource.node.addr, 'TagOption key', 1, 128, key);
-      tagOptionsList.forEach((value: string) => {
-        InputValidator.validateLength(resource.node.addr, 'TagOption value', 1, 256, value);
-        const tagOptionKey = hashValues(key, value, resourceStack.node.addr);
-        const tagOptionConstructId = `TagOption${tagOptionKey}`;
-        let cfnTagOption = resourceStack.node.tryFindChild(tagOptionConstructId) as CfnTagOption;
-        if (!cfnTagOption) {
-          cfnTagOption = new CfnTagOption(resourceStack, tagOptionConstructId, {
-            key: key,
-            value: value,
-            active: true,
-          });
-        }
-        const tagAssocationKey = hashValues(key, value, resource.node.addr);
-        const tagAssocationConstructId = `TagOptionAssociation${tagAssocationKey}`;
-        if (!resource.node.tryFindChild(tagAssocationConstructId)) {
-          new CfnTagOptionAssociation(resource as cdk.Resource, tagAssocationConstructId, {
-            resourceId: resourceId,
-            tagOptionId: cfnTagOption.ref,
-          });
-        }
-      });
-    };
+    for (const [tagOptionIdentifier, cfnTagOption] of Object.entries(tagOptions._tagOptionsMap)) {
+      const tagAssocationConstructId = `TagOptionAssociation${hashValues(resource.node.addr, tagOptionIdentifier)}`;
```

Comment:
I have been able to maintain current behavior here.  I didn't try to maintain this before because from my OOP life it felt wrong to add a more coupling between the two resources, and I more or less tracked the impact of changes that would happen as all or nothing - since we could expect to see the TagOption resources changing I thought it might just go hand in hand to expect associations to change as well.

If this new version is fine, my understanding is now better for it.